crates/common/src/compile.rs

```diff
@@ -271,9 +271,51 @@ impl ProjectCompiler {
     }
 }
 
-/// Map over artifacts contract sources name -> file_id -> (source, contract)
+/// Contract source code and bytecode.
 #[derive(Clone, Debug, Default)]
-pub struct ContractSources(pub HashMap<String, HashMap<u32, (String, ContractBytecodeSome)>>);
+pub struct ContractSources {
+    /// Map over artifacts contract sources name -> file_id -> (source, contract)
+    pub ids_by_name: HashMap<String, Vec<u32>>,
+    /// Map over artifacts contract sources file_id -> (source, contract)
+    pub sources_by_id: HashMap<u32, (String, ContractBytecodeSome)>,
+}
+
+impl ContractSources {
+    /// Inserts a contract into the sources.
+    pub fn insert(
+        &mut self,
+        artifact_id: &ArtifactId,
+        file_id: u32,
+        source: String,
+        bytecode: ContractBytecodeSome,
+    ) {
+        self.ids_by_name.entry(artifact_id.name.clone()).or_default().push(file_id);
+        self.sources_by_id.insert(file_id, (source, bytecode));
+    }
+
+    /// Returns the source for a contract by file ID.
+    pub fn get(&self, id: u32) -> Option<&(String, ContractBytecodeSome)> {
+        self.sources_by_id.get(&id)
+    }
+
+    /// Returns all sources for a contract by name.
+    pub fn get_sources(&self, name: &str) -> Option<Vec<(u32, &(String, ContractBytecodeSome))>> {
+        self.ids_by_name.get(name).map(|ids| {
+            ids.iter().filter_map(|id| Some((*id, self.sources_by_id.get(id)?))).collect()
+        })
+    }
+
+    /// Returns all (name, source) pairs.
+    pub fn entries(&self) -> Vec<(String, &(String, ContractBytecodeSome))> {
+        self.ids_by_name
+            .iter()
+            .flat_map(|(name, ids)| {
+                ids.iter()
+                    .filter_map(move |id| self.sources_by_id.get(id).map(|s| (name.clone(), s)))
+            })
+            .collect()
```

Comment:
Can return `impl Iterator` instead of collecting

Reply:
Was wrestling with types before, but think I got it now!

Reply:
Also made get_sources return an `impl Iterator`